### ISSUE TITLE
fluid迁移

### ISSUE THREAD

**Issue by AWSWXY** [PaddlePaddle/Paddle#17912]
v1和fluid 版本 效果差异比较大

**Comment by AWSWXY** [comment]
旧版本代码是v1 新版本是fluid版本，迁移过来之后效果差异比较大。
1、v1版本起始auc 在0.7 左右，fluid版本起始auc在0.5左右
2、fluid版本下降速度特别慢

**Comment by AWSWXY** [comment]
旧版本代码是v1 新版本是fluid版本，迁移过来之后效果差异比较大。
1、fluid版本下降速度特别慢

**Comment by qingqing01** [comment]
@AWSWXY  请注意，v1版本在backward传播梯度时gradient没有对batch size求平均。 在fluid里，如果超参数和v1都一致的情况下，相应的需要对loss做readuce_sum，而非求mean。 另外，其他数据处理、网络也检查需要保持一致。

**Comment by AWSWXY** [comment]
修改后无明显好转

**Comment by AWSWXY** [comment]
fluid版本网络


    def  mynet(user_fea_dim, query_fea_dim, use_py_reader=True, is_infer=False):
    EMBED_SIZE=128
    user_input_fea = fluid.layers.data(name='user_input_fea', shape=[1], dtype="int64", lod_level=1)
    query_input_fea = fluid.layers.data(name='query_input_fea', shape=[1], dtype="int64", lod_level=1)
    label = fluid.layers.data(name='label', shape=[1], dtype='int64')
    datas=[]
    datas.append(user_input_fea)
    datas.append(query_input_fea)
    datas.append(label) 
    words = datas
    py_reader = None
    if use_py_reader:
        py_reader = fluid.layers.create_py_reader_by_data(capacity=64,
                                                      feed_list=datas,
                                                      name="test_reader" if is_infer else "train_reader",
                                                      use_double_buffer=True)
        words = fluid.layers.read_file(py_reader)

    usr_emb_features = fluid.layers.embedding(
        input=words[0],
        size=[user_fea_dim, EMBED_SIZE],
        dtype='float32',
        param_attr=fluid.ParamAttr(
                        initializer=fluid.initializer.Normal(scale=1/math.sqrt(user_fea_dim))),
        is_sparse=True,
        is_distributed=False)

    query_emb_features = fluid.layers.embedding(
        input=words[1],
        size=[query_fea_dim, EMBED_SIZE],
        dtype='float32',
        param_attr=fluid.ParamAttr(
                        initializer=fluid.initializer.Normal(scale=1/math.sqrt(query_fea_dim))),
        is_sparse=True,
        is_distributed=False)

    usr_emb_features_hidden = fluid.layers.sequence_pool(
        input=usr_emb_features, pool_type="sum")

    query_emb_features_hidden = fluid.layers.sequence_pool(
        input=query_emb_features, pool_type="sum")

    usr_combined_features = fluid.layers.fc(input=usr_emb_features_hidden , size=50, act="tanh")
    query_combined_features = fluid.layers.fc(input=query_emb_features_hidden , size=50, act="tanh")

    
    usr_combined_features = fluid.layers.dropout(x=usr_combined_features, dropout_prob=0.3)

    inference = fluid.layers.cos_sim(X=usr_combined_features, Y=query_combined_features)

    scale_infer = fluid.layers.fc(input=inference, size=2, act='softmax')
    cost = fluid.layers.cross_entropy(input=scale_infer, label=words[2])
    #avg_cost = fluid.layers.mean(square_cost)
    avg_cost = fluid.layers.mean(cost)

    accuracy = fluid.layers.accuracy(input=scale_infer, label=words[2])
    auc_var, batch_auc_var, auc_states = \
        fluid.layers.auc(input=scale_infer, label=words[2], num_thresholds=2 ** 12, slide_steps=20)

    return avg_cost, accuracy, auc_var, batch_auc_var, scale_infer, py_reader, words


    optimizer = fluid.optimizer.RMSProp(learning_rate=0.1*1e-4)
    _, params_grads = optimizer.minimize(loss)
   batch_size = 1000

v1版本网络



    settings(
        batch_size = 1000,
        learning_rate=0.1*1e-4,
        #learning_method = AdamOptimizer(),
        #learning_method = AdaGradOptimizer(),
        learning_method=RMSPropOptimizer()
    )

    query_feature_layer = data_layer(name='query_feature', size=query_fea_dim)
    user_feature_layer = data_layer(name='user_feature', size=user_fea_dim)
    query_label = data_layer(name="query_label", size=1)
    query_hidden = fc_layer(input=query_feature_layer, size=50, act=TanhActivation())
    user_hidden = fc_layer(input=user_feature_layer, size=50, act=TanhActivation(),    
    layer_attr=ExtraLayerAttribute(drop_rate=0.3))
    prediction = cos_sim(a=query_hidden, b=user_hidden)
    if not is_predict:
        output = fc_layer(input=prediction, size=2, act=SoftmaxActivation())
        evaluator = auc_evaluator(output, query_label)
        outputs(classification_cost(input=output, label=query_label)) 
    else:
        outputs(user_hidden)
        outputs(query_hidden)

batch_size ，学习率，节点数已经对齐。

问题：
1、v1版本一个batch需要20分钟，fluid版本一个batch50分钟。
2、v1版本loss下降很快，fluid版本下降慢很多
3、auc几乎无上涨

**Comment by guru4elephant** [comment]
@AWSWXY 训练的是分布式版本还是单机版本？

**Comment by AWSWXY** [comment]
> @AWSWXY 训练的是分布式版本还是单机版本？

分布式版本

**Comment by AWSWXY** [comment]
经排查是网络结构问题，问题已经解决
